From fc784839c8a2cb3ea98098e2fdb9d6220df5b0db Mon Sep 17 00:00:00 2001
From: Meng Mingming <mingming.meng@nxp.com>
Date: Fri, 14 Apr 2017 16:41:04 +0800
Subject: [PATCH 8/8] MGS-2787 [#ccc] Check if g2d_format has an alpha when
 doing alpha blending

It need to check if g2d_format has an alpha when doing alpha blending.

Upstream Status: Inappropriate [i.MX specific]

Date: Apr 14, 2017
Signed-off-by: Meng Mingming <mingming.meng@nxp.com>
---
 src/g2d-renderer.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/g2d-renderer.c b/src/g2d-renderer.c
index e59838a..28c1c2d 100644
--- a/src/g2d-renderer.c
+++ b/src/g2d-renderer.c
@@ -365,6 +365,9 @@ g2d_SetSurfaceRect(struct g2d_surfaceEx* g2dSurface, g2dRECT* rect)
 	}
 }
 
+#define _hasAlpha(format) (format==G2D_RGBA8888 || format==G2D_BGRA8888 \
+    || format==G2D_ARGB8888 || format==G2D_ABGR8888)
+
 static int
 g2d_blitSurface(void *handle, struct g2d_surfaceEx * srcG2dSurface, struct g2d_surfaceEx *dstG2dSurface, 
 	g2dRECT *srcRect, g2dRECT *dstRect)
@@ -372,7 +375,11 @@ g2d_blitSurface(void *handle, struct g2d_surfaceEx * srcG2dSurface, struct g2d_s
 	g2d_SetSurfaceRect(srcG2dSurface, srcRect);
 	g2d_SetSurfaceRect(dstG2dSurface, dstRect);
 	srcG2dSurface->base.blendfunc = G2D_ONE;
-	dstG2dSurface->base.blendfunc = G2D_ONE_MINUS_SRC_ALPHA;
+    dstG2dSurface->base.blendfunc = G2D_ONE_MINUS_SRC_ALPHA;
+    if(!(_hasAlpha(srcG2dSurface->base.format)))
+    {
+        g2d_disable(handle, G2D_BLEND);
+    }
 
 	if(g2d_blitEx(handle, srcG2dSurface, dstG2dSurface))
 	{
-- 
2.9.3

